8.1.2/startup/generateStartupCommand.js

```diff
@@ -18,15 +18,45 @@ if (typeof process.env.WEBSITE_ROLE_INSTANCE_ID !== 'undefined'
     roleInstanceId = process.env.WEBSITE_ROLE_INSTANCE_ID;
 }
 
-var startupCommand = fs.readFileSync(CMDFILE, 'utf8').trim();
+// Is Application Insights enabled, with an associated ikey?
+var appInsightsEnabled = process.env.ENABLE_APPINSIGHTS && process.env.APPINSIGHTS_INSTRUMENTATIONKEY;
+var appInsightsPreloadArg = "--require /opt/startup/initAppInsights.js";
+var nodeCommandPrefix = "node ";
+
+if (appInsightsEnabled) {
+    console.log("Application Insights enabled");
+    nodeCommandPrefix += appInsightsPreloadArg + " ";
+}
+
+function augmentCommandIfNeccessary(command) {
+    if (!command || !appInsightsEnabled) {
+        return command;
+    }
+
+    // Application Insights is enabled, so we need to
+    // to update the specified startup command to pre-load it.
+    if (command.indexOf("pm2 start ") === 0) {
+        return command += " --node-args='" + appInsightsPreloadArg + "'";
+    } else if (command.indexOf("node ") === 0) {
+        // Simply replacing the prefix allows the user to specify
+        // additional Node flags, in addition to the AI preload one.
+        return command.replace("node ", nodeCommandPrefix);
+    }
+
+    // The command is using an unknown executable, and therefore,
+    // the App Insights runtime can't be automatically enabled.
+    return command;
+}
+
+var startupCommand = augmentCommandIfNeccessary(fs.readFileSync(CMDFILE, 'utf8').trim());
 
 // No user-provided startup command, check for scripts.start
 if (!startupCommand) {
     var packageJsonPath = "/home/site/wwwroot/package.json";
     var json = fs.existsSync(packageJsonPath) && JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'))
     if (typeof json == 'object' && typeof json.scripts == 'object' && typeof json.scripts.start == 'string') {
-        console.log("Found scripts.start in package.json")
-        startupCommand = 'npm start';
+        console.log("Found scripts.start in package.json");
+        startupCommand = augmentCommandIfNeccessary(json.scripts.start.trim());
```

Comment:
The only difference that I can think of is that `npm start` adds `./node_modules/.bin` to the process' `PATH`, and so it's possible that we could introduce issues for apps that depended on that behavior. However, that's typically only used to be able to run executables in the script that are actually installed as a local Node.js module (e.g. `webpack `), so I don't think this should be a problem in practice.

The bigger issue might be the assumption that the start script was running `node` in the first place, since you could technically put anything else there (e.g. `foobar -f server.js`).